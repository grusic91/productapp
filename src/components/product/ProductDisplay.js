import React from 'react';
import PropTypes from 'prop-types';
import { ProductEditor } from './ProductEditor';
import { ProductTable } from './ProductTable';

export class ProductDisplay extends React.Component {
    constructor(props){
        super(props);
        this.state = {
            showEditor: false,
            selectedProduct: null
        }
    }

    startEditing = product => {
        this.setState({
            showEditor: true,
            selectedProduct: product
        })
    }

    createProduct = () => {
        this.setState({
            showEditor: true,
            selectedProduct: {}
        })
    }

    cancelEditing = () => {
        this.setState({
            showEditor: false,
            selectedProduct: null
        })
    }

    saveProduct = product => {
        this.props.saveCallback(product);
        this.setState({
            showEditor: false,
            selectedProduct: null
        });
    }

    render() {
        const { showEditor, selectedProduct } = this.state;        

        return (
           <div className="m-2">            
            { showEditor ? 
                <div>
                    <ProductEditor
                        key={selectedProduct.id || -1 }
                        product={selectedProduct}
                        saveCallback={this.saveProduct}
                        cancelCallback={this.cancelEditing}
                />
                </div>
                    :
                <React.Fragment>
                    <ProductTable
                        products={this.props.products}
                        editCallback={this.startEditing}
                        deleteCallback={this.props.deleteCallback}
                    />
                    <div className="text-center">
                        <button className="btn btn-primary m-1" onClick={this.createProduct}>
                            Create Product
                        </button>
                    </div>
                </React.Fragment>                 
                }
            </div>      
        )
    }
}

ProductDisplay.propTypes = {
    name: PropTypes.string.isRequired,
    products: PropTypes.array.isRequired,
    saveCallback: PropTypes.func.isRequired,
    deleteCallback: PropTypes.func.isRequired,
}