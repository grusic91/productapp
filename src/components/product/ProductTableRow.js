import React from 'react';

export class ProductTableRow extends React.Component {
    render() {
        let p = this.props.product;
        if(p) {
            return <tr>
            <td>{ p.id }</td>
            <td>{ p.name }</td>
            <td>{ p.category}</td>
            <td className="text-right">${ Number(p.price).toFixed(2) }</td>
            <td>
                <button 
                    className="btn btn-sm btn-warning m-1"
                    onClick={ () => this.props.editCallback(p) }
                > 
                    Edit 
                </button>
                <button
                    className="btn btn-sm btn-danger m-1"
                    onClick={ () => this.props.deleteCallback(p) }
                >
                    Delete
                </button>
            </td>
        </tr>
        }
        
        return <h1>Loading...</h1>
    }
}
