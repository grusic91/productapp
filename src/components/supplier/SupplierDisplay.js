import React from 'react';
import PropTypes from 'prop-types';
import { SupplierEditor } from './SupplierEditor';
import { SupplierTable } from './SupplierTable';

export class SupplierDisplay extends React.Component {
    constructor(props){
        super(props);
        this.state = {
            showEditor: false,
            selected: null
        }
    }

    startEditing = supplier => {
        this.setState({
            showEditor: true,
            selected: supplier
        })
    }

    createSupplier = () => {
        this.setState({
            showEditor: true,
            selected: {}
        })
    }

    cancelEditing = () => {
        this.setState({
            showEditor: false,
            selected: null
        })
    }

    saveSupplier = supplier => {
        this.props.saveCallback(supplier);
        this.setState({
            showEditor: false,
            selected: null
        });
    }

    render() {
        const { showEditor, selected } = this.state;

        return (
            <div className="m-2">
            { showEditor ? 
                <div>
                    <SupplierEditor
                        key={selected.id || -1 }
                        supplier={selected}
                        saveCallback={this.saveSupplier}
                        cancelEditing={this.cancelEditing}
                    />
                </div>
                    :
                <React.Fragment>
                    <SupplierTable
                        suppliers={this.props.suppliers}
                        editCallback={this.startEditing}
                        deleteCallback={this.props.deleteCallback}
                    />
                    <div className="text-center">
                        <button className="btn btn-primary m-1" onClick={this.createSupplier}>
                            Create Supplier
                        </button>
                    </div>
                </React.Fragment>
            }
            </div>
        )
    }
}


SupplierDisplay.propTypes = {
    name: PropTypes.string.isRequired,
    suppliers: PropTypes.array.isRequired,
    saveCallback: PropTypes.func.isRequired,
    deleteCallback: PropTypes.func.isRequired,
}