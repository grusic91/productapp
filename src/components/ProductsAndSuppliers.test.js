import React from 'react';
import { shallow, mount } from 'enzyme';
import ProductsAndSuppliers from './ProductsAndSuppliers';
import { ProductDisplay } from './product/ProductDisplay';
import { SupplierDisplay } from './supplier/SupplierDisplay';
import { ProductEditor } from './product/ProductEditor';

describe("<ProductsAndSuppliers />", () => {
  let wrapper;

  beforeAll(() => wrapper = shallow(<ProductsAndSuppliers />));

  it("renders correctly", () => {
    expect(wrapper).toMatchSnapshot()
  });

  it("should render one Selector", () => {
    expect(wrapper.find('Selector')).toHaveLength(1);
  });

  it("should render the ProductDisplay and SupplierDisplay Components with required props", () => {
    const productDisplayProps = wrapper.find(ProductDisplay).props();
    const supplierDisplayProps = wrapper.find(SupplierDisplay).props();
    
    expect(wrapper.containsAllMatchingElements([
      <ProductDisplay {...productDisplayProps}/>,
      <SupplierDisplay {...supplierDisplayProps} />
    ])).toEqual(true);    
  });
});

