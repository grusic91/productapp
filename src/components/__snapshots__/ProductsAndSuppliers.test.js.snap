// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductsAndSuppliers /> renders correctly 1`] = `ShallowWrapper {}`;
